multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java

```diff
@@ -169,30 +181,53 @@ public ResponseEntity<AsyncUploadResult> getUploadFromUrlJob(String spaceGuid, S
             return ResponseEntity.notFound()
                                  .build();
         }
+        return getAsyncUploadResult(spaceGuid, namespace, job);
+    }
 
+    private ResponseEntity<AsyncUploadResult> getAsyncUploadResult(String spaceGuid, String namespace, AsyncUploadJobEntry job) {
         if (job.getState() == State.RUNNING || job.getState() == State.INITIAL) {
-            var count = jobCounters.getOrDefault(jobId, new AtomicLong(-1));
-            return ResponseEntity.ok(ImmutableAsyncUploadResult.builder()
-                                                               .status(AsyncUploadResult.JobStatus.RUNNING)
-                                                               .bytes(count.get())
-                                                               .build());
+            Future<?> futureTask = runningTasks.get(job.getId());
+            if (futureTask == null) {
+                LOGGER.error(MessageFormat.format(Messages.JOB_0_WAS_NOT_FOUND_IN_THE_RUNNING_TASKS, job.getId()));
+                return ResponseEntity.ok(createErrorResult(Messages.JOB_IS_NOT_BEING_EXECUTED,
+                                                           AsyncUploadResult.ClientAction.RETRY_UPLOAD));
+            }
+            if (!futureTask.isDone()) {
+                var count = jobCounters.getOrDefault(job.getId(), new AtomicLong(-1));
+                return ResponseEntity.ok(ImmutableAsyncUploadResult.builder()
+                                                                   .status(AsyncUploadResult.JobStatus.RUNNING)
+                                                                   .bytes(count.get())
+                                                                   .build());
+            }
+            var jobWithLatestState = getJob(job.getId(), spaceGuid, namespace);
```

Comment:
The idea is to fetch the new state of the object from the database as it should be changed because the Future task is done. This covers the race condition between updating/getting the job
I do not think the variable can be null, we've already checked it